third-party/spamassassin/src/main/java/org/apache/james/spamassassin/SpamAssassinInvoker.java

```diff
@@ -102,56 +103,56 @@ public SpamAssassinResult scanMail(MimeMessage message, Username username) throw
                 .sneakyThrow());
     }
 
-    public SpamAssassinResult scanMail(MimeMessage message) throws MessagingException {
-        return metricFactory.runPublishingTimerMetric(
+    public Mono<SpamAssassinResult> scanMail(MimeMessage message) {
+        return Mono.from(metricFactory.runPublishingTimerMetric(
             "spamAssassin-check",
-            Throwing.supplier(
-                () -> scanMailWithoutAdditionalHeaders(message))
-            .sneakyThrow());
+            scanMailWithoutAdditionalHeaders(message)));
     }
 
-    private SpamAssassinResult scanMailWithAdditionalHeaders(MimeMessage message, String... additionalHeaders) throws MessagingException {
-        try (Socket socket = new Socket(spamdHost, spamdPort);
-             OutputStream out = socket.getOutputStream();
-             BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(out);
-             PrintWriter writer = new PrintWriter(bufferedOutputStream);
-             BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
-
-            LOGGER.debug("Sending email {} for spam check", message.getMessageID());
-
-            writer.write("CHECK SPAMC/1.2");
-            writer.write(CRLF);
-
-            Arrays.stream(additionalHeaders)
-                .forEach(header -> {
-                    writer.write(header);
-                    writer.write(CRLF);
-                });
-
-            writer.write(CRLF);
-            writer.flush();
-
-            // pass the message to spamd
-            message.writeTo(out);
-            out.flush();
-            socket.shutdownOutput();
-
-            SpamAssassinResult spamAssassinResult = in.lines()
-                .filter(this::isSpam)
-                .map(this::processSpam)
-                .findFirst()
-                .orElse(SpamAssassinResult.empty());
-
-            LOGGER.debug("spam check result: {}", spamAssassinResult);
-            return spamAssassinResult;
-        } catch (UnknownHostException e) {
-            throw new MessagingException("Error communicating with spamd. Unknown host: " + spamdHost);
-        } catch (IOException | MessagingException e) {
-            throw new MessagingException("Error communicating with spamd on " + spamdHost + ":" + spamdPort, e);
-        }
+    private Mono<SpamAssassinResult> scanMailWithAdditionalHeaders(MimeMessage message, String... additionalHeaders) {
+       return Mono.fromCallable( () -> {
+            try (Socket socket = new Socket(spamdHost, spamdPort);
```

Comment:
Can't we use reactor-netty?

Reply:
i didn't tough of that, it's a good suggestion

Reply:
but the reactor netty API doesn't really allow to do this in a simple way.

Reply:
Don't the TCP client allows outbound string/stream sending, and receiving a flux of byte buffer we can put in a stream + reader + & do the magic with almost the code that is here?

Reply:
Not that i know, the http api of reactor netty is nice. The TCP one less so : 
https://projectreactor.io/docs/netty/release/reference/index.html#_consuming_data_2

Reply:
did you look at https://github.com/reactor/reactor-netty/blob/master/src/test/java/reactor/netty/tcp/TcpClientTests.java#L218 ?